src/Compilers/CSharp/Portable/Symbols/ReducedExtensionMethodSymbol.cs

```diff
@@ -156,6 +162,66 @@ private static MethodSymbol InferExtensionMethodTypeArguments(MethodSymbol metho
                 return null;
             }
 
+            // For the purpose of constraint checks we use error type symbol in place of type arguments that we couldn't infer from the first argument.
+            // This prevents constraint checking from failing for corresponding type parameters.
+            int firstNullInTypeArgs = -1;
+            var notInferredTypeParameters = PooledHashSet<TypeParameterSymbol>.GetInstance();
+            var typeParams = method.TypeParameters;
+            var typeArgsForConstraintsCheck = typeArgs;
+            for (int i = 0; i < typeArgsForConstraintsCheck.Length; i++)
+            {
+                if (!typeArgsForConstraintsCheck[i].HasType)
+                {
+                    firstNullInTypeArgs = i;
+                    var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+                    builder.AddRange(typeArgsForConstraintsCheck, firstNullInTypeArgs);
+
+                    for (; i < typeArgsForConstraintsCheck.Length; i++)
+                    {
+                        var typeArg = typeArgsForConstraintsCheck[i];
+                        if (!typeArg.HasType)
+                        {
+                            notInferredTypeParameters.Add(typeParams[i]);
+                            builder.Add(TypeWithAnnotations.Create(ErrorTypeSymbol.UnknownResultType));
+                        }
+                        else
+                        {
+                            builder.Add(typeArg);
+                        }
+                    }
+
+                    typeArgsForConstraintsCheck = builder.ToImmutableAndFree();
+                    break;
+                }
+            }
```

Comment:
Could this loop be simplified, perhaps by checking for missing types up front?
```C#
PooledHashSet<TypeParameterSymbol> notInferredTypeParameters = null;
if (typeArgs.Any(t => !t.HasType))
{
    var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
    notInferredTypeParameters = PooledHashSet<TypeParameterSymbol>.GetInstance();
    for (int i = 0; i < typeArgs.Length; i++)
    {
        ...
    }
    typeArgsForConstraintsCheck = builder.ToImmutableAndFree();
}
```

Reply:
Perhaps, but I literally just restored this code that was removed in a previous PR.  I'd rather minimize the scope for error (I would like to get it into 16.3) by keeping it identical.

---
In reply to: [315453057](https://github.com/dotnet/roslyn/pull/38109#discussion_r315453057) [](ancestors = 315453057)